### PR TITLE
Follwer

### PR DESCRIPTION
**Pull request by jackie174**
done follower part, done post and post with authentication, done like with GET method,
For like with POST method, we need first work on Inbox, right now for GET method is done

### CODE REVIEW

File: backend/backend/urls.py [hunk 1]
```diff
@@ -13,20 +13,24 @@
...
     path('service/authors/<str:author_id>/posts/', include("post.urls")),
     path('service/authors/<str:author_id>/posts/<str:post_id>', include("post.urls")),
     path('service/', include(('backend.routers', 'backend'), namespace='backend-api')),
     path('service/authors/<str:author_id>/posts/<str:post_id>/comments/', include("comments.urls")),
-
-
+    path('service/authors/<str:author_id>/followers/', include("followers.urls")),
+    path('service/authors/<str:author_id>/friends/<str:foreign_author_id>', TrueFriendApiView.as_view(), name = "check if true friends"),
+    path('service/authors/<str:author_id>/posts/<str:post_id>/likes', LikesPostApiView.as_view(), name="post like"),
+    path('service/authors/<str:author_id>/likes/', LikesAuthorApiView.as_view(), name="author like"),
```
Comment (nervuna) [inline backend/backend/urls.py]: Should change 'service/authors/<str:author_id>/likes/' to 'service/authors/<str:author_id>/liked/' in accordance to spec, and to help differentiate between likes that are in a post and the posts that the author "liked".
```suggestion
    path('service/authors/<str:author_id>/liked/', LikesAuthorApiView.as_view(), name="author like"),
```

File: backend/followers/views.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+class FollowersApiView(GenericAPIView):
+    """
+    URL: ://service/authors/{AUTHOR_ID}/followers
+    GET [local, remote]: get a list of authors who are AUTHOR_ID’s followers
+    """
+    permission_classes = [IsAuthenticated]
+    serializer_class = FollowerSerializer
+    def get(self, request, author_id):
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
```
Comment (nervuna) [inline backend/followers/views.py]: We shouldn't need to check if the requester is the author, since this is a public method. Anyone should be able to go to
'://service/authors/{AUTHOR_ID}/followers' without needing authentication.
```suggestion
```

File: backend/followers/views.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+from post.serializers import PostSerializer
+from author.serializers import AuthorSerializer, FollowerSerializer
+from post.views import check_author_id, get_author_id, get_foreign_id, get_friend_id
+
+class FollowersApiView(GenericAPIView):
+    """
+    URL: ://service/authors/{AUTHOR_ID}/followers
+    GET [local, remote]: get a list of authors who are AUTHOR_ID’s followers
+    """
+    permission_classes = [IsAuthenticated]
```
Comment (nervuna) [inline backend/followers/views.py]: See below comment.
```suggestion
    permission_classes = [AllowAny]
```

File: backend/followers/views.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+from post.models import Post
+from author.models import Author
+from rest_framework import response, status
+from rest_framework.generics import GenericAPIView
+from rest_framework.authentication import BasicAuthentication
+from rest_framework.permissions import IsAuthenticated
```
Comment (nervuna) [inline backend/followers/views.py]: See below comments.
```suggestion
from rest_framework.permissions import IsAuthenticated, AllowAny, IsAuthenticatedOrReadOnly
```

File: backend/followers/views.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                    return response.Response(followers_serializer_list, status=status.HTTP_200_OK)
+                else:
+                    followers_serializer_list = {
+                    "type": "followers",
+                    "items": []
+                    }
+                    return response.Response(followers_serializer_list, status=status.HTTP_200_OK)
+
+            except Exception as e:
+                return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)
```
Comment (nervuna) [inline backend/followers/views.py]: Again, GET should be a public method and shouldn't need authentication.
```suggestion
    permission_classes = [IsAuthenticatedOrReadOnly]
    serializer_class = FollowerSerializer
    def get(self, request, author_id, foreign_author_id):
          try:
              author_id = get_author_id(request)
              foreign_id = get_foreign_id(request)
              
              current_author = Author.objects.get(id = author_id)
              followers = current_author.followers.all()
              if followers.exists():
                  followers = current_author.followers.all().order_by('displayName')
                  followers_serializer = self.serializer_class(followers, many=True)
                  followers_serializer_list = {
                      "type": "followers",
                      "items": followers_serializer.data
                  }
                  return response.Response(followers_serializer_list, status=status.HTTP_200_OK)
              else:
                  followers_serializer_list = {
                  "type": "followers",
                  "items": []
                  }
                  return response.Response(followers_serializer_list, status=status.HTTP_200_OK)

          except Exception as e:
              return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)
```

File: backend/followers/views.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    def put(self, request, author_id, foreign_author_id):
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
+        else:
+            try:
+                author_id = get_author_id(request)
```
Comment (nervuna) [inline backend/followers/views.py]: 1. put() takes in a parameter `author_id` but you replace it with `author_id = get_author_id(request)`. These variables should either have different names to avoid confusion or you don't do `get_author_id(request)`.
2. Your check_author_id() method is faulty:
```python
def check_author_id(request):
    author_id= get_author_id(request)
    author = Author.objects.filter(id = author_id)
    return author.exists()
```
   It checks if the author requested exists, but if this is the case then the HTTP response shouldn't be 401, it should be 404 NOT FOUND. 
If you really want to check if the requester is the actual author that they are requesting to change, you should check the JWT payload (what they claim to be), though, this is already done by IsAuthenticated and the default JWT authentication class I added in settings.py. Additionally, IsAuthenticated will already return a 401 if they're not authenticated. 
To summarize the above: either remove the `if check_author_id(request) == False:` clause, or, return a 404 in that if clause.

File: backend/followers/views.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                    return response.Response("Error: Foreign author do not found", status=status.HTTP_404_NOT_FOUND)
+               
+                return response.Response(result, status=status.HTTP_200_OK)
+
+            except Exception as e:
+                return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)
+    def delete(self, request, author_id, foreign_author_id):
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
+        else:
```
Comment (nervuna) [inline backend/followers/views.py]: Same as the above comment.

File: backend/followers/views.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+class TrueFriendApiView(GenericAPIView):
+    """
+    URL: ://service/authors/{AUTHOR_ID}/friends
+    GET [local, remote] get all true friend of AUTHOR_ID
+    """
+    permission_classes = [IsAuthenticated]
+    serializer_class = FollowerSerializer
+    def get(self, request, author_id, foreign_author_id):
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
```
Comment (nervuna) [inline backend/followers/views.py]: Public method, don't need authentication.
```suggestion
    permission_classes = [AllowAny]
    serializer_class = FollowerSerializer
    def get(self, request, author_id, foreign_author_id):
        if check_author_id(request) == False:
            return response.Response(status=status.HTTP_404_NOT_FOUND)
```

File: backend/like/views.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            author = Author.objects.get(id = author_id)
+            try:
+            
+                post_like = Like.objects.filter(author = author)
+                post_likes = self.serializer_class(post_like, many=True)
+                result = {"type": "liked", "items": post_likes.data}
+                return response.Response(result, status=status.HTTP_200_OK)
+            except Exception as e:
+                return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)
+
```
Comment (nervuna) [inline backend/like/views.py]: All the methods for Likes are public, no authentication is needed. 
```suggestion
from rest_framework.permissions import AllowAny
from rest_framework import response, status
from post.views import get_author_id, get_post_id, check_author_id
from post.models import Post
from author.models import Author
from like.models import Like
from django.db.models import Q
from like.Serializers import LikePostSerializer, LikeAuthorSerializer
# Create your views here.


class LikesPostApiView(GenericAPIView):
    permission_classes = [AllowAny]
    serializer_class=LikePostSerializer
    def get(self, request, author_id, post_id):
        post_id = get_post_id(request)
        if check_author_id(request) == False:
            return response.Response(status=status.HTTP_404_NOT_FOUND)
        else:
            author_id = get_author_id(request)
            author = Author.objects.get(id = author_id)
            try:
                post = Post.objects.get(id = post_id, author=author)
                if post ==None:
                    return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)
                post_like = Like.objects.filter(object = post.id)
                post_likes = self.serializer_class(post_like, many=True)
                result = {"type": "likes", "items": post_likes.data}
                return response.Response(result, status=status.HTTP_200_OK)
            except Exception as e:
                return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)


class LikesAuthorApiView(GenericAPIView):
    permission_classes = [AllowAny]
    serializer_class=LikeAuthorSerializer
    def get(self, request, author_id):
        if check_author_id(request) == False:
            return response.Response(status=status.HTTP_404_NOT_FOUND)
        else:
            author_id = get_author_id(request)
            author = Author.objects.get(id = author_id)
            try:
            
                post_like = Like.objects.filter(author = author)
                post_likes = self.serializer_class(post_like, many=True)
                result = {"type": "liked", "items": post_likes.data}
                return response.Response(result, status=status.HTTP_200_OK)
            except Exception as e:
                return response.Response(f"Error: {e}", status=status.HTTP_404_NOT_FOUND)

```

File: backend/like/views.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    permission_classes = [IsAuthenticated]
+    serializer_class=LikeAuthorSerializer
+    def get(self, request, author_id):
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
+        else:
+            author_id = get_author_id(request)
+            author = Author.objects.get(id = author_id)
+            try:
+
```
Comment (nervuna) [inline backend/like/views.py]: This statement should be in the try clause, because what if the author is not found? an exception is raised but there's nothing to catch it.

```suggestion
            
            try:
                author = Author.objects.get(id = author_id)
```

File: backend/like/views.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    serializer_class=LikePostSerializer
+    def get(self, request, author_id, post_id):
+        post_id = get_post_id(request)
+        if check_author_id(request) == False:
+            return response.Response(status=status.HTTP_401_UNAUTHORIZED)
+        else:
+            author_id = get_author_id(request)
+            author = Author.objects.get(id = author_id)
+            try:
+                post = Post.objects.get(id = post_id, author=author)
```
Comment (nervuna) [inline backend/like/views.py]: See below.
```suggestion
            try:
                author = Author.objects.get(id = author_id)
                post = Post.objects.get(id = post_id, author=author)
```

File: backend/post/views.py [hunk 12]
```diff
@@ -167,21 +200,45 @@ def get_author_id(request):
...
+    else:
+        xx=request.build_absolute_uri().split('service/')
+        author_id= xx[0]+xx[1]
+        return author_id
+
 def check_author_id(request):
     author_id= get_author_id(request)
     author = Author.objects.filter(id = author_id)
-    print(author.exists())
     return author.exists()
```
Comment (nervuna) [inline backend/post/views.py:243]: A couple of things to note here, but I'll leave it as is. I'll make the changes to post once you make your other changes.

**Review by nervuna — CHANGES_REQUESTED** [review]
Once you make your changes, I'll merge it to main.

**Review by nervuna — APPROVED** [review]
👍

File: backend/followers/views.py (hunk 6 above)
Comment (jackie174) [reply backend/followers/views.py]: I do get_author_id beacuase we need to get host from request,
This time, I change to get_author_url_id(request), which is get host with author_id,
